# 4.9 Task Data Visualization Part_1

# 1. Import analysis libraries

In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

# 2. Import dataframes and check the dimensions

In [2]:
# Create path variable

path= r'C:\Users\babum\Desktop\CF_Data Analytics\Data Immersion\Achievement 4\02-2023 Instacart Basket Analysis'

In [3]:
# Import customers dataframe

df_cust = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
# Check the output

df_cust.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
df_cust.shape

(206209, 10)

In [6]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [7]:
# Import prepared dataframe

ords_prods_merge = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data', 'orders_products_grouped.pkl'))

In [8]:
# Check the ouput

ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...    price_range_loc     busiest_day         busiest_days  \
0  ...  Mid-range product  Regularly busy  Regularly busy days   
1  ...  Mid-range product  Regularly busy  Regularly busy days   
2  ...  Mid-range product     Busiest day         Busiest days   
3  ...  Mid-range product  Regularly busy         Slowest days   
4  ...  Mid-range product      Least busy         Slowest days   

  busiest_period_of_day max_order      loyalty_flag avg_price spending_flag  \
0           Most orders        32  Regular customer  6.935811   Low spender   
1        Average orders        32  Regular customer  6.935811   Low spender   
2        Average orders         5      New customer  7.930208   Low spender   
3           Most orders         3      New customer  4.972414   Low spender   
4        Average orders         3      New customer  4.972414   Low spender   

   median_days_prior_order order_frequency_flag  
0                      8.0    Frequent customer  
1                      8.0    Frequent customer  
2                      8.0    Frequent customer  
3                      9.0    Frequent customer  
4                      9.0    Frequent customer  

[5 rows x 24 columns]

In [9]:
ords_prods_merge.shape

(32404859, 24)

# 3. Data wrangling

# Renaming a column

In [10]:
df_cust.columns

Index(['user_id', 'First Name', 'Surnam', 'Gender', 'STATE', 'Age',
       'date_joined', 'n_dependants', 'fam_status', 'income'],
      dtype='object')

In [11]:
#Rename a column

df_cust.rename(columns = {'First Name' : 'first_name','Surnam' : 'surname','Gender' : 'gender','STATE' : 'state','Age' : 'age','n_dependants' : 'num_of_dependants','fam_status' : 'family_status'}, inplace = True)

# The column names of First Name, Surnam, Gender, STATE, Age, n_dependants and fam_status has been renamed for consistent logic.

In [12]:
# Check the output

df_cust.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_of_dependants family_status  income  
0                  3       married  165665  
1                  0        single   59285  
2                  2       married   99568  
3                  0        single   42049  
4                  1       married   40374

# Changing the datatype

In [13]:
# Change the datatype of columns to solve memory usage issues

df_cust ['user_id']= df_cust ['user_id'].astype('int32')
df_cust ['age']= df_cust ['age'].astype('int8')
df_cust ['num_of_dependants']= df_cust ['num_of_dependants'].astype('int8')
df_cust ['income']= df_cust ['income'].astype('int32')

# The datatype of 'user_id', 'age', 'num_of_dependants', and 'income' has been changed to solve the memory usage issues.

In [14]:
df_cust.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column             Non-Null Count   Dtype 
---  ------             --------------   ----- 
 0   user_id            206209 non-null  int32 
 1   first_name         194950 non-null  object
 2   surname            206209 non-null  object
 3   gender             206209 non-null  object
 4   state              206209 non-null  object
 5   age                206209 non-null  int8  
 6   date_joined        206209 non-null  object
 7   num_of_dependants  206209 non-null  int8  
 8   family_status      206209 non-null  object
 9   income             206209 non-null  int32 
dtypes: int32(2), int8(2), object(6)
memory usage: 11.4+ MB


# Check for mixed datatype

In [15]:
#Check for mixed type data

for col in df_cust.columns.tolist():
    weird = (df_cust[[col]].applymap(type) != df_cust[[col]].iloc[0].apply(type)).any(axis = 1)
if len(df_cust[weird])>0:
    print(col)

# There are no mixed data type

# 4. Data consistency 

# Check for missing values

In [16]:
# Finding missing values

df_cust.isnull().sum()

user_id                  0
first_name           11259
surname                  0
gender                   0
state                    0
age                      0
date_joined              0
num_of_dependants        0
family_status            0
income                   0
dtype: int64

In [17]:
#Subset only the missing values

df_nan = df_cust[df_cust['first_name'].isnull() == True]

In [18]:
df_nan

user_id first_name   surname  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        num_of_dependants family_status  income  
53                      2       married   41709  
73                      0        single   82518  
82                      3       married  117099  
99                      1       married  155673  
105                     3       married  151819  
...                   ...           ...     ...  
206038                  3       married   87783  
206044                  2       married  108488  
206090                  1       married   45275  
206162                  0        single   56325  
206171                  2       married   59222  

[11259 rows x 10 columns]

In [19]:
#Drop column first_name and surname

df_cust_new = df_cust.drop(columns = ['first_name','surname'])

# The first name column contains 11,259 missing values. Since this column is not used in the analysis 'first_name' and 'surname' columns are dropped and a new dataframe was created.

In [20]:
df_cust_new.head()

user_id  gender       state  age date_joined  num_of_dependants  \
0    26711  Female    Missouri   48    1/1/2017                  3   
1    33890  Female  New Mexico   36    1/1/2017                  0   
2    65803    Male       Idaho   35    1/1/2017                  2   
3   125935  Female        Iowa   40    1/1/2017                  0   
4   130797  Female    Maryland   26    1/1/2017                  1   

  family_status  income  
0       married  165665  
1        single   59285  
2       married   99568  
3        single   42049  
4       married   40374

# Finding duplicates

In [21]:
#Finding duplicates

df_dups = df_cust_new[df_cust_new.duplicated()]

In [22]:
df_dups.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 0 entries
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   user_id            0 non-null      int32 
 1   gender             0 non-null      object
 2   state              0 non-null      object
 3   age                0 non-null      int8  
 4   date_joined        0 non-null      object
 5   num_of_dependants  0 non-null      int8  
 6   family_status      0 non-null      object
 7   income             0 non-null      int32 
dtypes: int32(2), int8(2), object(4)
memory usage: 0.0+ bytes


# There are no duplicates in the data

# 5. Combining two dataframes

In [23]:
ords_prods_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'price_range_loc', 'busiest_day', 'busiest_days',
       'busiest_period_of_day', 'max_order', 'loyalty_flag', 'avg_price',
       'spending_flag', 'median_days_prior_order', 'order_frequency_flag'],
      dtype='object')

In [24]:
#Drop column "_merge"

ords_prods_merge = ords_prods_merge.drop(columns = ['_merge'])

In [25]:
#Merging data

ords_prods_all = ords_prods_merge.merge(df_cust_new, on = ['user_id'], how = 'left', indicator = True) 

In [26]:
ords_prods_all.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   ...  median_days_prior_order  order_frequency_flag  gender      state age  \
0  ...                      8.0     Frequent customer    Male  Minnesota  81   
1  ...                      8.0     Frequent customer    Male  Minnesota  81   
2  ...                      8.0     Frequent customer  Female    Vermont  66   
3  ...                      9.0     Frequent customer  Female  Wisconsin  40   
4  ...                      9.0     Frequent customer  Female  Wisconsin  40   

  date_joined num_of_dependants  family_status  income  _merge  
0    8/1/2019                 1        married   49620    both  
1    8/1/2019                 1        married   49620    both  
2   6/16/2018                 2        married  158302    both  
3    2/9/2020                 3        married   31308    both  
4    2/9/2020                 3        married   31308    both  

[5 rows x 31 columns]

In [27]:
ords_prods_all['_merge'].value_counts(dropna = False)

both          32404859
left_only            0
right_only           0
Name: _merge, dtype: int64

In [28]:
ords_prods_all.shape

(32404859, 31)

In [29]:
ords_prods_all.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 31 columns):
 #   Column                   Dtype   
---  ------                   -----   
 0   product_id               int32   
 1   product_name             object  
 2   aisle_id                 int32   
 3   department_id            int32   
 4   prices                   float64 
 5   order_id                 int32   
 6   user_id                  int32   
 7   order_number             int8    
 8   orders_day_of_week       int8    
 9   order_hour_of_day        int8    
 10  days_since_prior_order   float16 
 11  add_to_cart_order        int32   
 12  reordered                int8    
 13  price_range_loc          object  
 14  busiest_day              object  
 15  busiest_days             object  
 16  busiest_period_of_day    object  
 17  max_order                int32   
 18  loyalty_flag             object  
 19  avg_price                float64 
 20  spending_flag         

In [30]:
ords_prods_all.isnull().sum()

product_id                       0
product_name                     0
aisle_id                         0
department_id                    0
prices                        5127
order_id                         0
user_id                          0
order_number                     0
orders_day_of_week               0
order_hour_of_day                0
days_since_prior_order     2076096
add_to_cart_order                0
reordered                        0
price_range_loc                  0
busiest_day                      0
busiest_days                     0
busiest_period_of_day            0
max_order                        0
loyalty_flag                     0
avg_price                        0
spending_flag                    0
median_days_prior_order          5
order_frequency_flag             5
gender                           0
state                            0
age                              0
date_joined                      0
num_of_dependants                0
family_status       

# 6. Export dataframe

In [31]:
#Export df to pkl format

ords_prods_all.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_all.pkl'))